File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.util.Queue;
import java.util.Scanner;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;

public class Main {
    static int[][] map;
    static boolean[][] visited;
    static int num=2;
    static int n;
    static int[] dx={0,1,0,-1};
    static int[] dy={1,0,-1,0};
    static ArrayList<Integer> apt=new ArrayList<>();
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n=sc.nextInt();
        map=new int[n][n];
        visited=new boolean[n][n];
        int cnt=0;
        for(int i=0;i<n;i++){
            String[] temp=sc.next().split("");
            for(int j=0;j<n;j++){
                map[i][j]=Integer.parseInt(temp[j]);
            }
        }
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                if(map[i][j]==1 && !visited[i][j]){
                    apt.add(BFS(i,j));
                    cnt++;
                    num+=1;
                }
            }
        }

        Collections.sort(apt);
        System.out.println(cnt);
        for(int i=0;i<apt.size();i++){
            System.out.println(apt.get(i));
        }
    }

    private static int BFS(int i, int j) {
        Queue<int[]> queue=new LinkedList<>();
        queue.offer(new int[] {i,j});
        visited[i][j]=true;
        int result=1;
        while(!queue.isEmpty()){
            int[] now=queue.poll();
            for(int k=0;k<4;k++){
                int x=now[0]+dx[k];
                int y=now[1]+dy[k];
                if(x>=0&&x<n&&y>=0&&y<n){
                    if(!visited[x][y]){
                        visited[x][y]=true;
                        if(map[x][y]!=0){
                            map[x][y]=num;
                            result++;
                            queue.add(new int[] {x,y});
                        }
                    }
                }
            }
        }
        return result;
    }
}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
from collections import deque
from pkgutil import resolve_name

N = int(input())

board = [[int(num) for num in input()] for _ in range(N)]
result = [[0 for _ in range(N)] for _ in range(N)]
visited = [[False for _ in range(N)] for _ in range(N)]
houseCount = list()

queue = deque()
houseNumber = 1
dx = [0,1,0,-1]
dy = [1,0,-1,0]

def bfs():
    count = 1
    while queue:
        x, y = queue.popleft()
        for i in range(4):
            nx, ny = x+dx[i], y+dy[i]

            if nx>=0 and nx<N and ny<N and ny>=0 and not visited[nx][ny]:
                visited[nx][ny]=True
                if board[nx][ny] == 1:
                    result[nx][ny] = houseNumber
                    queue.append([nx,ny])
                    count+=1

    return count
for i in range(N):
    for j in range(N):
        if not visited[i][j] and board[i][j] == 1:
            queue.append([i,j])
            visited[i][j]= True
            houseCount.append(bfs())
            houseNumber+=1
houseCount.sort()
print(len(houseCount))
for num in houseCount:
    print(num)
